libzeth/snarks/groth16/groth16_snark.tcc

```diff
@@ -45,232 +44,104 @@ template<typename ppT>
+typename groth16_snark<ppT>::ProvingKeyT groth16_snark<
+    ppT>::proving_key_read_bytes(std::istream &is)
 {
-    if (path.empty()) {
-        // Used for a debugging purpose
-        boost::filesystem::path tmp_path = get_path_to_debug_directory();
-        boost::filesystem::path proof_json_file("proof.json");
-        path = tmp_path / proof_json_file;
+    ProvingKeyT pk;
+    is >> pk;
+    if (!is_well_formed<ppT>(pk)) {
+        throw std::invalid_argument("proving key (read) not well-formed");
     }
-    // Convert the boost path into char*
-    const char *str_path = path.string().c_str();
-
-    std::stringstream ss;
-    std::ofstream fh;
-    fh.open(str_path, std::ios::binary);
-
-    ss << "{"
-       << "\n"
-       << "\t\"a\" :[" << point_g1_affine_to_hexadecimal_str<ppT>(proof.g_A)
-       << "],"
-       << "\n"
-       << "\t\"b\" :[" << point_g2_affine_to_hexadecimal_str<ppT>(proof.g_B)
-       << "],"
-       << "\n"
-       << "\t\"c\" :[" << point_g1_affine_to_hexadecimal_str<ppT>(proof.g_C)
-       << "],"
-       << "\n"
-       << "}";
+    return pk;
+}
 
-    ss.rdbuf()->pubseekpos(0, std::ios_base::out);
-    fh << ss.rdbuf();
-    fh.flush();
-    fh.close();
+template<typename ppT>
+std::ostream &groth16_snark<ppT>::keypair_write_bytes(
+    std::ostream &os, const typename groth16_snark<ppT>::KeypairT &keypair)
+{
+    proving_key_write_bytes(keypair.pk, os);
+    verification_key_write_bytes(keypair.vk, os);
+    return os;
 }
 
 template<typename ppT>
-void groth16_snark<ppT>::write_keypair(
-    std::ostream &out, const typename groth16_snark<ppT>::KeypairT &keypair)
+typename groth16_snark<ppT>::KeypairT groth16_snark<ppT>::keypair_read_bytes(
+    std::istream &is)
 {
-    if (!is_well_formed<ppT>(keypair.pk)) {
-        throw std::invalid_argument("proving key (write) not well-formed");
-    }
-    if (!is_well_formed<ppT>(keypair.vk)) {
-        throw std::invalid_argument("verification key (write) not well-formed");
-    }
-    out << keypair.pk;
-    out << keypair.vk;
+    ProvingKeyT pk = proving_key_read_bytes(is);
+    VerificationKeyT vk = verification_key_read_bytes(is);
+    return libsnark::r1cs_gg_ppzksnark_keypair<ppT>(
+        std::move(pk), std::move(vk));
 }
 
 template<typename ppT>
-typename groth16_snark<ppT>::KeypairT groth16_snark<ppT>::read_keypair(
-    std::istream &in)
+std::ostream &groth16_snark<ppT>::proof_write_json(
+    const typename groth16_snark<ppT>::ProofT &proof, std::ostream &os)
 {
-    libsnark::r1cs_gg_ppzksnark_keypair<ppT> keypair;
-    in >> keypair.pk;
-    in >> keypair.vk;
-    if (!is_well_formed<ppT>(keypair.pk)) {
-        throw std::invalid_argument("proving key (read) not well-formed");
-    }
-    if (!is_well_formed<ppT>(keypair.vk)) {
-        throw std::invalid_argument("verification key (read) not well-formed");
-    }
-    return keypair;
+    os << "{\n"
+       << "    \"a\" :[" << point_g1_affine_to_json<ppT>(proof.g_A) << "],\n"
```

Comment:
Do you need the `:["` and `"],` around `point_g1_affine_to_json` here? It seems that the square brackets are already set here : https://github.com/clearmatics/zeth/blob/clean-libzeth-conventions/libzeth/core/group_element_utils.tcc#L18

Reply:
(Same question for the rest of the files and files in pghr13)

Reply:
I missed these.  Thanks.  I beleive they are all fixed now.